src/main/java/com/google/sps/servlets/utils/Statistics.java

```diff
@@ -14,29 +14,83 @@
 
 package com.google.sps.servlets.utils;
 
-import com.google.sps.servlets.Range;
+import java.math.BigDecimal;
+import java.util.Comparator;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
 
 public class Statistics {
-  // Aggregated score for a list of comments
+  private static final double INTERVAL = 0.2;
+  private static final double UPPER_END = 1.0;
+  private static final double LOWER_END = -1.0;
+  // Contains sentiment scores in the range [-1, 1] with given intervals.
   private Map<Range, Integer> aggregateValues;
   private double averageScore;
 
+  public Map<Range, Integer> getAggregateValues() {
+    return aggregateValues;
+  }
+
+  public double getAverageScore() {
+    return averageScore;
+  }
+
   public Statistics(List<Double> sentimentScores) {
+    sentimentScores =
+        sentimentScores.stream()
+            .filter(score -> (score >= LOWER_END && score <= UPPER_END))
+            .collect(Collectors.toList());
     setAggregateScores(sentimentScores);
     setAverageScore(sentimentScores);
   }
 
+  /**
+   * Categorize all score values into different range intervals and count the frequency for each
+   * interval, and set the aggregatedValues.
+   *
+   * @param sentimentScores a list of score values from -1.0 to 1.0
+   */
   private void setAggregateScores(List<Double> sentimentScores) {
-    // TODO(Xin): Add sorting code
     aggregateValues = new HashMap<>();
-    aggregateValues.put(new Range(-1.0, 1), 1);
+    BigDecimal curPoint = BigDecimal.valueOf(LOWER_END);
+    BigDecimal interval = BigDecimal.valueOf(INTERVAL);
+
+    // Initialize the HashMap with intervals
+    while (curPoint.doubleValue() < UPPER_END) {
+      Range currentRange =
+          new Range(
+              curPoint.doubleValue(), Math.min(curPoint.add(interval).doubleValue(), UPPER_END));
+      curPoint = curPoint.add(interval);
+      aggregateValues.put(currentRange, 0);
+    }
+    // Add score's interval to different ranges two sorting with and two pointers pop-up
+    sentimentScores.sort(Comparator.naturalOrder());
+    curPoint = BigDecimal.valueOf(LOWER_END);
+    int scoreIdx = 0;
+    while ((scoreIdx < sentimentScores.size()) && (curPoint.doubleValue() < UPPER_END)) {
```

Comment:
Why not just for loop here, you know you have filtered out scores that are out of bounds.
I am also confused about why we are making new foundRange all the time, when we kinda already made them above.
Maybe consider a double for loop with sorted range as the outer loop and sentimentScores as the inner loop. You can have an index variable set outside of the double for loop to keep track of the index in sentimentScores so the double for loop still gives you O(n).